next/src/visibility.ts

```diff
@@ -0,0 +1,164 @@
+ * @param values - The current form values
+ * @param schema - The JSON schema definition
+ * @param options - Validation options
+ */
+export function updateFieldVisibility(
+  fields: Field[],
+  values: SchemaValue,
+  schema: JsfObjectSchema,
+  options: ValidationOptions = {},
+) {
+  if (!isObjectValue(values)) {
+    return
+  }
+
+  // Apply rules to current level of fields
+  applySchemaRules(fields, values, schema, options)
+
+  // Process nested object fields that have conditional logic
+  for (const fieldName in schema.properties) {
+    const fieldSchema = schema.properties[fieldName]
+
+    // Only process object schemas with conditional logic (allOf)
+    if (typeof fieldSchema !== 'object' || fieldSchema === null
+      || Array.isArray(fieldSchema) || !fieldSchema.allOf) {
+      continue
+    }
+
+    const objectField = fields.find(f => f.name === fieldSchema.title || f.name === fieldName)
+    if (!objectField || !objectField.fields || objectField.fields.length === 0) {
+      continue
+    }
+
+    const fieldValues = isObjectValue(values[fieldName])
+      ? values[fieldName]
+      : isObjectValue(values[objectField.name]) ? values[objectField.name] : {}
+
+    // Apply rules to nested fields
+    applySchemaRules(objectField.fields, fieldValues, fieldSchema as JsfObjectSchema, options)
+
+    // Only process nested fields if parent is visible
+    if (objectField.isVisible) {
+      updateFieldVisibility(
+        objectField.fields,
+        fieldValues,
+        fieldSchema as JsfObjectSchema,
+        options,
+      )
+    }
+  }
+}
+
+/**
+ * Applies JSON Schema conditional rules to determine field visibility
+ * @param fields - The fields to apply rules to
+ * @param values - The current form values
+ * @param schema - The JSON schema containing the rules
+ * @param options - Validation options
+ *
+ * Fields start with visibility based on their required status.
+ * Conditional rules in the schema's allOf property can then:
+ * - Make fields visible by including them in a required array
+ * - Make fields hidden by setting them to false in properties
+ */
+function applySchemaRules(
+  fields: Field[],
+  values: SchemaValue,
+  schema: JsfObjectSchema,
+  options: ValidationOptions = {},
+) {
+  if (!schema.allOf || !Array.isArray(schema.allOf) || !isObjectValue(values)) {
+    return
+  }
+
+  const conditionalRules = schema.allOf
+    .filter(rule => typeof rule === 'object' && rule !== null && 'if' in rule)
+    .map((rule) => {
+      const ruleObj = rule as NonBooleanJsfSchema
+
+      const ifErrors = validateSchema(values, ruleObj.if!, options)
+      const matches = ifErrors.length === 0
+
+      // Prevent fields from being shown when required fields have type errors
+      let hasTypeErrors = false
+      if (matches
+        && typeof ruleObj.if === 'object'
+        && ruleObj.if !== null
+        && 'required' in ruleObj.if
+        && Array.isArray(ruleObj.if.required)) {
```

Comment:
Do we need both condition? Isn't the `Array.isArray` enough?

```suggestion
        && Array.isArray(ruleObj.if.required)) {
```

Reply:
Great catch, yes the double check here is redundant.
I think even `matches && Array.isArray(ruleObj?.if?.required)` should work? I'll try that.

Reply:
Update:
> I think even `matches && Array.isArray(ruleObj?.if?.required)` should work?

Does not work because we must still verify that `ruleObj.if` is an actual object because it could be a boolean. So I've just removed the redundant `'required' in ruleObj.if` check.